tests/end-to-end.rs

```diff
@@ -161,28 +167,54 @@ where
 
 
 /// Serve our registry.
-fn serve_registry() -> (JoinHandle<()>, PathBuf, SocketAddr) {
-  let root = tempdir().unwrap();
-  let path = root.path().to_owned();
+fn serve_registry(root_path: RegistryRootPath) -> (JoinHandle<()>, PathBuf, SocketAddr) {
+  let path = match root_path {
+    RegistryRootPath::Absolute => tempdir().unwrap().path().to_owned(),
+    RegistryRootPath::Relative => {
+      let root = tempdir().unwrap().into_path();
+      let base_dir = root.parent().unwrap();
+      env::set_current_dir(base_dir).unwrap();
```

Comment:
Why do we need to change the current working directory everywhere? This is a nightmare in a multi-threaded context. Plus, if I remove the call everything is fine...

Reply:
This is needed to make passing relative directory to `Index::new()` work inside the temporary directory. If you remove this then it the index root will be re-created by the `Index::new()` in the directory where you run the test command, i.e. project root dir most likely

Reply:
I don't get it. I don't see anything in `Index::new` relying on the current working directory. If there is something, that's probably a bug. But sprinkling `set_current_dir` here is not a fix. Again, tests run concurrently and will trample over each other's `set_current_dir` calls.

Reply:
As soon as you pass a relative path in `root` argument to `Index::new()` then everything starts depending on the current directory :) E.g. if we pass a literal `"index-root"` as `root` argument then for example `Index::ensure_config()` will write `config.json` to `"index-root/config.json"` which effectively writes to `"<CWD>/index-root/config.json"`.

I understand your concerns, but I don't really see an easier way to test that relative paths work while still remaining constrained to the temporary directory.

Reply:
Ultimately I don't mind dropping this test.

Reply:
It sounds as if `Index` should be performing the conversion to absolute path internally then and work with that, no?

Reply:
For that to work we still need to change the current directory into the temporary area before calling `fs::canonicalize()`, otherwise it will resolve to a directory under project root instead of temp dir.

Reply:
Well, technically you don't need inherent support for async. You can always start a reactor manually and `spawn()` an async task.

What should we do with the PR for now? If the main logic if fine for you I would prefer to merge it with this test dropped/commented.

Reply:
> For that to work we still need to change the current directory into the temporary area before calling `fs::canonicalize()`, otherwise it will resolve to a directory under project root instead of temp dir.

Yeah, I realized that as well. But I still think we should be able to pass in a suitably crafted relative path. E.g., if `cargo-http-registry` is at `/home/user/cargo-http-registry` and `tempdir` creates `/tmp/.tmpYb5HBA/`, the relative path should be `../../../tmp/.tmpYb5HBA/`. I would guess that would be sufficient for testing? Or do we specifically need something without parent directory involvement?

Reply:
I though about this, but was hesitating to bring another dep (e.g. `pathdiff`). Additionally I think it won't always conceptually work on Windows if temp dir and project dir are on the different drives (e.g. C: and D:).

Reply:
The changes in `index.rs` don't really affect anything now that I changed all invocations of `Index::add` to only use relative paths. If you roll back both `index.rs` and `publish.rs` changes you'll get failure in `publish_relative_index_root()` test.

Reply:
The test now uses `pathdiff` to calculate the relative path without setting the current working directory.